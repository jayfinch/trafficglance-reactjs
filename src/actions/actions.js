import type from './constants'
import fetchJsonp from 'fetch-jsonp'
import * as commutesHelper from '../utils/commutes-helper'
import * as trafficHelper from '../utils/traffic-helper'

function parseResponse (response) {
  if (!response.ok) {
    throw Error(response.statusText)
  }
  return response.json()
}

// fetchConfig

export function fetchConfig (uri = '/config.json') {
  return function (dispatch) {
    dispatch(fetchConfigRequest())

    return fetch(uri)
    .then(response => parseResponse(response))
    .then(json => dispatch(fetchConfigSuccess(json)))
    .catch(e => dispatch(fetchConfigFailure(e.message)))
  }
}
export function fetchConfigRequest () {
  return {
    type: type.FETCH_CONFIG_REQUEST
  }
}

export function fetchConfigSuccess (data) {
  return {
    type: type.FETCH_CONFIG_SUCCESS,
    apiKey: data.apiKey,
    distanceUnit: data.distanceUnit,
    commutes: commutesHelper.transformCommutes(data.commutes)
  }
}

export function fetchConfigFailure (message) {
  return {
    type: type.FETCH_CONFIG_FAILURE,
    error: message
  }
}

// fetchTraffic

export function fetchTraffic (url, id) {
  return function (dispatch) {
    dispatch(fetchTrafficRequest(id))

    return fetchJsonp(url, { jsonpCallback: 'jsonp' })
    .then(response => parseResponse(response))
    .then(json => dispatch(fetchTrafficSuccess(id, json)))
    .catch(e => dispatch(fetchTrafficFailure(id, e.message)))
  }
}

export function fetchTrafficRequest (id) {
  return {
    type: type.FETCH_TRAFFIC_REQUEST,
    id: id
  }
}

export function fetchTrafficSuccess (id, bingData) {
  const subTree = trafficHelper.getSubTree(bingData)

  return {
    type: type.FETCH_TRAFFIC_SUCCESS,
    id: id,
    trafficData: {
      ...trafficHelper.getDurationTime(subTree),
      ...trafficHelper.getDurationByCongestion(subTree),
      distance: trafficHelper.getDistance(subTree),
      arriveTime: trafficHelper.getArriveTime(subTree)
    }
  }
}

export function fetchTrafficFailure (id, message) {
  return {
    type: type.FETCH_TRAFFIC_FAILURE,
    id: id,
    error: message
  }
}
