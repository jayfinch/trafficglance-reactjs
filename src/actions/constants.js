export default {
  REQUEST_CONFIG: 'request-config',
  RECEIVE_CONFIG: 'receive-config',
  FAIL_CONFIG: 'fail-config',
  REQUEST_TRAFFIC: 'request-traffic',
  RECEIVE_TRAFFIC: 'receive-traffic',
  FAIL_TRAFFIC: 'fail-traffic'
}
