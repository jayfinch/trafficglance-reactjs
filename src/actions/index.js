import type from './constants'
import fetchJsonp from 'fetch-jsonp'

export function requestConfig () {
  return {
    type: type.REQUEST_CONFIG
  }
}

export function receiveConfig (data) {
  return {
    type: type.RECEIVE_CONFIG,
    data: data
  }
}

export function failConfig (error) {
  return {
    type: type.FAIL_CONFIG,
    error: error
  }
}

export function requestTraffic (id) {
  return {
    type: type.REQUEST_TRAFFIC,
    id: id,
    isFetching: false
  }
}

export function receiveTraffic (id, trafficData) {
  return {
    type: type.RECEIVE_TRAFFIC,
    id: id,
    trafficData: trafficData
  }
}

export function failTraffic (id, error) {
  return {
    type: type.FAIL_TRAFFIC,
    id: id,
    error: error
  }
}

// async

export function fetchConfig () {
  return function (dispatch) {
    dispatch(requestConfig())

    return fetch('/config.json')
    .then((response) => response.json())
    .then((json) => dispatch(receiveConfig(json)))
    .catch((e) => dispatch(failConfig(e)))
  }
}

export function fetchTraffic (url, id) {
  return function (dispatch) {
    dispatch(requestTraffic(id))

    return fetchJsonp(url, {
      jsonpCallback: 'jsonp'
    })
    .then((response) => response.json())
    .then((json) => {
      if (json.statusCode === 200) {
        dispatch(receiveTraffic(id, json))
      } else {
        throw new Error('Problem with Bing data')
      }
    })
    .catch((e) => dispatch(failTraffic(id, e)))
  }
}
