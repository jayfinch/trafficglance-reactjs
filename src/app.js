import React from 'react'
import ReactDOM from 'react-dom'
import Route from './route.js'

var TrafficGlanceApp = React.createClass({
  propTypes: {
    data: React.PropTypes.array,
    units: React.PropTypes.string
  },

  _renderRoutes: function () {
    var units = this.props.units
    return this.props.data.map(function (route, i) {
      return (
        <Route name={route.name} url={route.url} key={i} units={units} />
      )
    })
  },

  render: function () {
    return (
      <div>
        <div className='toolbar'>
          <a href='#'><img src='images/refresh-circle.svg' className='refresh-all' /></a>
          <div className='site-name'>TrafficGlance</div>
        </div>
        <div className='container'>
          <div className='row' id='routes'>
            {this._renderRoutes()}
          </div>
          <div className='row credit'>
            <div className='col-xs-12'><a href='#'>Powered by Bingâ„¢</a></div>
          </div>
        </div>
      </div>
    )
  }
})

module.exports = TrafficGlanceApp
