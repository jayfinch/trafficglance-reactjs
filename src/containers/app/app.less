@import (reference) "../../styles/constants.less";

// bootstrap overrides

body {
  background: #f2f2f2;
  color: #333;
  font-size: 13px;
  line-height: 1.2;
}

.container {
  width: auto;
  max-width: 1200px;
  margin-top: @toolbar-height + 30px;;
  padding: 0 15px;

  @media (min-width: 768px) {
    padding: 0 30px;
  }
}

// wrap every 2 items

@media (min-width: 768px) and (max-width: 991px) {
  .route {
    float: left;
  }
  .route:nth-child(2n+1) {
    clear: left;
  }
}

// wrap every 3 items

@media (min-width: 992px) {
  .route {
    float: left;
  }
  .route:nth-child(3n+1) {
    clear: left;
  }
}
