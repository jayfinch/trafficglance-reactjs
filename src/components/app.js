import React, { Component, PropTypes } from 'react'
import Commute from './commute'
import { bindActionCreators } from 'redux'
import { connect } from 'react-redux'
import * as actions from '../actions'
import { buildUrl } from '../utils/bing'
import 'whatwg-fetch'

class App extends Component {

  componentDidMount () {
    this.props.actions.fetchConfig()
  }

  fetchCommuteTraffic (i, segments) {
    const { distanceUnit, apiKey, actions } = this.props
    const url = buildUrl(segments, distanceUnit, apiKey)
    actions.fetchTraffic(url, i)
  }

  onClickRefreshAll (event) {
    event.preventDefault()
    this.props.commutes.forEach((commute, i) => {
      this.fetchCommuteTraffic(i, commute.segments)
    })
  }

  render () {
    const { commutes, distanceUnit, configError } = this.props

    var commutesEl = <div className='loading'>Loading...</div>

    if (configError) {
      commutesEl = <div className='loading'>You must set up your config</div>
    }

    if (commutes.length) {
      commutesEl = commutes.map((commute, i) => {
        return (
        <Commute
          key={i}
          name={commute.name}
          url={commute.url}
          units={distanceUnit}
          fetchingTraffic={commute.fetchingTraffic}
          fetchTraffic={() => this.fetchCommuteTraffic(i, commute.segments)}
          trafficData={commute.trafficData}
          trafficError={commute.trafficError }/>
        )
      })
    }

    return (
      <div>
        <div className='toolbar'>
          <div className='site-name'>
            TrafficGlance
          </div>
          <a href='#' className='refresh-all' onClick={this.onClickRefreshAll.bind(this)}>
            <span className='glyphicon glyphicon-refresh' aria-hidden='true'></span>
          </a>
        </div>
        <div className='container'>
          <div className='row' id='routes'>
            {commutesEl}
          </div>
          <div className='row credit'>
            <div className='col-xs-12'>
              Powered by Bingâ„¢
            </div>
          </div>
        </div>
      </div>
    )
  }
}

App.propTypes = {
  actions: PropTypes.object.isRequired,
  commutes: PropTypes.arrayOf(PropTypes.shape({
    id: PropTypes.number.isRequired,
    name: PropTypes.string,
    url: PropTypes.string.isRequired,
    fetchingTraffic: PropTypes.bool,
    trafficError: PropTypes.bool,
    segments: PropTypes.arrayOf(PropTypes.shape({
      waypointType: PropTypes.string.isRequired,
      latitude: PropTypes.string.isRequired,
      longitude: PropTypes.string.isRequired
    })),
    trafficData: PropTypes.object
  })).isRequired,
  distanceUnit: PropTypes.oneOf(['mi', 'km']).isRequired,
  apiKey: PropTypes.string.isRequired,
  configError: PropTypes.bool
}

function mapStateToProps (state) {
  return {...state}
}

function mapDispatchToProps (dispatch) {
  return {
    actions: bindActionCreators(actions, dispatch)
  }
}

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(App)
