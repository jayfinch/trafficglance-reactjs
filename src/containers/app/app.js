import React, { Component, PropTypes } from 'react'
import { bindActionCreators } from 'redux'
import { connect } from 'react-redux'
import * as actions from '../../actions/actions'
import { getApiUrl } from '../../helpers/bing-helper'
import Commutes from '../../components/commutes/commutes'
import Toolbar from '../../components/toolbar/toolbar'
import Footer from '../../components/footer/footer'
import './app.less'

export class App extends Component {

  constructor (props) {
    super(props)
    this.onClickRefresh = this.onClickRefresh.bind(this)
    this.onClickRefreshAll = this.onClickRefreshAll.bind(this)
  }

  componentDidMount () {
    this.props.actions.fetchConfig()
  }

  onClickRefresh (i, segments) {
    this.fetchCommuteTraffic(i, segments)
  }

  onClickRefreshAll () {
    this.props.commutes.forEach((commute, i) => {
      this.fetchCommuteTraffic(i, commute.segments)
    })
  }

  fetchCommuteTraffic (i, segments) {
    const { distanceUnit, apiKey, actions } = this.props
    const url = getApiUrl(segments, distanceUnit, apiKey)
    actions.fetchTraffic(url, i)
  }

  render () {
    const {
      commutes,
      distanceUnit,
      configError
    } = this.props

    return <div>
      <Toolbar
        onClickRefreshAll={this.onClickRefreshAll}
        isRefreshAvailable={Boolean(commutes.length)}
      />
      <div className='container'>
        <Commutes
          commutes={commutes}
          distanceUnit={distanceUnit}
          configError={configError}
          onClickRefresh={this.onClickRefresh}
        />
        <Footer />
      </div>
    </div>
  }
}

App.propTypes = {
  actions: PropTypes.object.isRequired,
  commutes: PropTypes.array,
  distanceUnit: PropTypes.oneOf(['mi', 'km']).isRequired,
  apiKey: PropTypes.string.isRequired,
  configError: PropTypes.bool
}

export function mapStateToProps (state) {
  return {...state}
}

export function mapDispatchToProps (dispatch) {
  return {
    actions: bindActionCreators(actions, dispatch)
  }
}

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(App)
