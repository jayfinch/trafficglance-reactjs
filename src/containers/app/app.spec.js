import React from 'react'
import { bindActionCreators } from 'redux'
import * as actions from '../../actions/actions'
import { App, mapStateToProps, mapDispatchToProps } from './app'
import renderer from 'react-test-renderer'

jest.mock('redux', () => ({
  bindActionCreators: jest.fn(() => ({}))
}))

jest.mock('../../helpers/bing-helper', () => {
  return { getApiUrl: jest.fn(() => 'mock-url') }
})

jest.mock('../../actions/actions', () => {
  return { fetchCommuteTraffic: jest.fn() }
})

describe('<App />', () => {
  describe('render', () => {
    it('Should render', () => {
      let tree = renderer.create(
        <App
          actions={{
            fetchConfig: jest.fn(),
            fetchTraffic: jest.fn()
          }}
          commutes={[]}
          distanceUnit='mi'
          apiKey='apikey'
          configError
        />
      ).toJSON()

      expect(tree).toMatchSnapshot()
    })
  })

  describe('onClickRefresh', () => {
    it('Should call fetchTraffic', () => {
      let instance = null
      const mockActions = {
        fetchConfig: jest.fn(),
        fetchTraffic: jest.fn()
      }

      renderer.create(<App
        actions={mockActions}
        commutes={[]}
        distanceUnit='mi'
        apiKey='apikey'
        configError={false}
        ref={(ref) => { instance = ref }}
      />)

      instance.onClickRefresh(7, [])
      expect(mockActions.fetchTraffic).toHaveBeenCalledWith('mock-url', 7)
    })
  })

  describe('onClickRefreshAll', () => {
    it('Should call fetchTraffic', () => {
      let instance = null
      const mockActions = {
        fetchConfig: jest.fn(),
        fetchTraffic: jest.fn()
      }

      renderer.create(<App
        actions={mockActions}
        commutes={[
          {
            id: 0,
            name: 'foo',
            url: 'bar',
            fetchingTraffic: false,
            segments: []
          },
          {
            id: 1,
            name: 'a',
            url: 'b',
            fetchingTraffic: false,
            segments: []
          }
        ]}
        distanceUnit='mi'
        apiKey='apikey'
        configError={false}
        ref={(ref) => { instance = ref }}
      />)

      instance.onClickRefreshAll()
      expect(mockActions.fetchTraffic).toHaveBeenCalledWith('mock-url', 0)
      expect(mockActions.fetchTraffic).toHaveBeenCalledWith('mock-url', 1)
    })
  })

  describe('mapStateToProps', () => {
    it('Should return new but identical object', () => {
      let state = { someProp: 'foo' }
      let actual = mapStateToProps(state)

      expect(actual).toEqual(state)
      expect(actual === state).toEqual(false)
    })
  })

  describe('mapDispatchToProps', () => {
    it('Should return bound actions', () => {
      let dispatch = jest.fn
      let actual = mapDispatchToProps(dispatch)

      expect(bindActionCreators).toHaveBeenCalledWith(actions, dispatch)
      expect(actual.actions).toBeTruthy()
    })
  })
})
