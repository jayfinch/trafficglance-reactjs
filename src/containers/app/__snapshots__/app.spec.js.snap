exports[`<App /> render Should render 1`] = `
<div>
  <div
    className="toolbar">
    <div
      className="toolbar__site-name">
      TrafficGlance
    </div>
  </div>
  <div
    className="container">
    <div
      className="row"
      id="routes">
      <div
        className="loading">
        You must set up your config
      </div>
    </div>
    <div
      className="row footer">
      <div
        className="col-xs-12">
        Powered by Bingâ„¢
      </div>
    </div>
  </div>
</div>
`;
